eRFC: single-file packages ("cargo script") integration (#3424)
rust-lang/rfcs: text/3424-cargo-script.md

- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

This *experimental RFC* adds unstable support for single-file
packages in cargo so we can explore the design and resolve
questions with an implementation to collect feedback on.

Single-file packages are `.rs` files with an embedded
manifest.  These will be accepted with just like `Cargo.toml` files with
`--manifest-path`.  `cargo` will be modified to accept `cargo <file>.rs` as a
shortcut to `cargo run --manifest-path <file>.rs`.  This allows placing
`cargo` in a `#!` line for directly running these files.

Example:
```rust
#!/usr/bin/env cargo

//! ```cargo
//! [dependencies]
//! clap = { version = "4.2", features = ["derive"] }
//! ```

use clap::Parser;

#[derive(Parser, Debug)]
#[clap(version)]
struct Args {
    #[clap(short, long, help = "Path to config")]
    config: Option<std::path::PathBuf>,
}

fn main() {
    let args = Args::parse();
    println!("{:?}", args);
}
```
```console

> $ ./prog --config file.toml

Reviewer: Does this console invocation need to have the `.rs` extension on the filename?
```suggestion
$ ./prog.rs --config file.toml
```
I was following along earlier in the pre-rfc thread on IRLO, but wasn't sure whether the `.rs` extension was still an unresolved question. (I see the other `hello_world.rs` example does have the `.rs` filename extension)

Author: > Does this console invocation need to have the .rs extension on the filename?

Either one is fine here.  For the utility workflows, people don't always like to put extensions on their scripts (independent of the language). 

Granted, that can affect what we do for `cargo test --manifest-path ./prog`.  We might not be able to key just off of the extension.

> I was following along earlier in the pre-rfc thread on IRLO, but wasn't sure whether the .rs extension was still an unresolved question. (I see the other hello_world.rs example does have the .rs filename extension)

Most of that extension discussion is about what extension would be used when it is present and not so much on the use cases for when to drop it.

Reviewer 2: > Either one is fine here.

How is this supposed to work in general? If I name the file `run`, then cargo will be invoked as `cargo run`, which looks like the subcommand and not like a request to execute a file. `cargo <file>.rs` is at least unambiguous.

Author: The exact precedence rules are left to be explored with the implementation.  This is recorded in "Unresolved questions" as

> If cargo, what precedence does it have compared to built-in commands, aliases, and external commands.

In epage/cargo-script-mvs#154, I've started looking into prior art among other build tools that support `<tool> <file>`.

dlang's `dub` is of particular interest
1. If the `<positional>` is `-` or ends with `.rs`, treat it as a single-file package
2. If `<positional>` matches an existing subcommand, run it
3. If `<positional>` exists, treat it as a single-file package
4. If `<positional>.d` exists, treat it as a single-file package
5. Otherwise, treat `<positional>` as a non-existent subcommand

For me, some tweaks I can see making
- Don't support "`<positional>.d` exists" check (4)
- For (1), check if `<positional>` has path separators (at least on my system, `#!` causes an absolute path to be passed to the command)

Reviewer 3: > * For (1), check if `<positional>` has path separators (at least on my system, `#!` causes an absolute path to be passed to the command)

it doesn't on my system (Android).

Reviewer 4: Anything involving paths is likely to be system specific. Or at least they'll be edge case systems.

Reviewer 5: > Either one is fine here. For the utility workflows, people don't always like to put extensions on their scripts (independent of the language).

People can just use a symlink though. In the case of the PHP PHAR format, PHP denies loading phar files that do not have a `.phar` extension for some reason (https://github.com/php/php-src/blob/a65cdd97a148dccdb68a5f58c258d54c305f1b2e/ext/phar/phar.c#L1961), so some installations of composer are performed by `ln -s composer.phar /usr/bin/composer` (citation needed: I know several people who do so, but I am not sure if this is a common practice).

Reviewer 6: > If I name the file `run`, then cargo will be invoked as `cargo run`, which looks like the subcommand and not like a request to execute a file.

`cargo -- run`?

Does this work?

 `#!/usr/bin/env cargo -- --`

D:
- [dub](https://dub.pm/advanced_usage)
  - `dub hello.d` is shorthand for `dub run --single hello.d`
  - Regular nested block comment (not doc-comment) at top of file with `dub.sdl:` header

Java
- [JEP 330: Launch Single-File Source-Code Programs](https://openjdk.org/jeps/330)
- [jbang](https://www.jbang.dev/)
  - `jbang init` w/ templates
  - `jbang edit` support, setting up a recommended editor w/ environment
  - Discourages `#!` and instead encourages looking like shell code with `///usr/bin/env jbang "$0" "$@" ; exit $?`
  - Dependencies and compiler flags controlled via comment-directives, including
    - `//DEPS info.picocli:picocli:4.5.0` (gradle-style locators)
      - Can declare one dependency as the source of versions for other dependencies (bom-pom)
    - `//COMPILE_OPTIONS <flags>`
    - `//NATIVE_OPTIONS <flags>`
    - `//RUNTIME_OPTIONS <flags>`
  - Can run code blocks from markdown
  - `--code` flag to execute code on the command-line
  - Accepts scripts from `stdin`

Haskell
- [`runghc` / `runhaskell`](https://downloads.haskell.org/ghc/latest/docs/users_guide/runghc.html)
  - Users can use the file stem (ie leave off the extension) when passing it in

> - [cabal's single-file haskel script](https://cabal.readthedocs.io/en/stable/getting-started.html#run-a-single-file-haskell-script)

Reviewer 7: typo here

```diff
-haskel
+haskell
```

- Feature Name: cargo-script
- Start Date: 2023-04-26
- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

This *experimental RFC* adds unstable support for single-file
packages in cargo so we can explore the design and resolve
questions with an implementation to collect feedback on.

Single-file packages are `.rs` files with an embedded
manifest.  These will be accepted with just like `Cargo.toml` files with
`--manifest-path`.  `cargo` will be modified to accept `cargo <file>.rs` as a
shortcut to `cargo run --manifest-path <file>.rs`.  This allows placing
`cargo` in a `#!` line for directly running these files.

Example:
```rust
#!/usr/bin/env cargo

> //! ```cargo
> //! [dependencies]
> //! clap = { version = "4.2", features = ["derive"] }
> //! ```

Reviewer 8: Does this need to be a doc comment? Could it just be an ordinary comment? If so, I think it should be, to make it more obvious that this isn't magic syntax.

```suggestion
// ```cargo
// [dependencies]
// clap = { version = "4.2", features = ["derive"] }
// ```
```

Reviewer 8: The reference-level explanation should explain that this can be a regular `//` comment, a `///` doc comment, or a `//!` file-level doc comment.

Author: The strawman examples in the eRFC are using doc-comments.  I've called out how we store the embedded manifest as one of the unresolved questions that we plan to resolve before stablization.  During the Pre-RFC, we [explored a lot of options](https://github.com/epage/cargo-script-mvs/blob/main/0000-cargo-script.md#embedded-manifest-format).  I left these out of the eRFC to avoid getting side tracked by things that aren't meant to be resolved at this time.

Reviewer 8: @epage We don't have to specify all of that now; I'd just like to make sure that people don't assume this *has* to be a doc comment.

Author: The current proposal is for it to be doc comments (an alternative was to instead accept regular comments). @Muscraft is prototyping the initial implementation of this proposal and has switched to `syn` to extract the doc comments, finding that our regex parser for the doc comments was accepting cases that aren't turned into `#![doc]`s.

I've clarified this within the guide section.

Reviewer 9: Wouldn't it be better to have specific syntax, or a macro, instead?

Whenever languages have comments which are parsed and have effects, it feels a bit uneasy. Things which are commented out should never have any effect, I think.

But maybe this is different, since I suppose rustc really does ignore these comments.

Author: For more of an idea of what syntaxes we'll consider, see the Pre-RFC. We decided to do an eRFC, rather than a full RFC, and only included a rough sketch to give people an idea of what the experiment will look like, deferring actual decisions to the experiment and approval of the experiment's result.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

As an eRFC, this is meant to convey what we are looking to
accomplish.  Many of the details may change before
stablization.

### Creating a New Package

*(Adapted from [the cargo book](https://doc.rust-lang.org/cargo/guide/creating-a-new-project.html))*

To start a new [package][def-package] with Cargo, create a file named `hello_world.rs`:
```rust
#!/usr/bin/env cargo

fn main() {
    println!("Hello, world!");
}
```

Let's run it
```console
$ chmod +x hello_world.rs
$ ./hello_world.rs
Hello, world!
```

### Dependencies

*(Adapted from [the cargo book](https://doc.rust-lang.org/cargo/guide/dependencies.html))*

[crates.io] is the Rust community's central [*package registry*][def-package-registry]
that serves as a location to discover and download
[packages][def-package]. `cargo` is configured to use it by default to find
requested packages.

#### Adding a dependency

To depend on a library hosted on [crates.io], you modify `hello_world.rs`:
```rust
#!/usr/bin/env cargo

//! ```cargo
//! [dependencies]
//! time = "0.1.12"
//! ```

fn main() {

>     println!("Hello, world!");

Reviewer 8: Nit, but the code here in this example should actually use the added dependency.

Author: (another concerned raised is how stable the version numbers are)

}
```

The `cargo` section is called a [***manifest***][def-manifest], and it contains all of the
metadata that Cargo needs to compile your package. This is written in the
[TOML] format (pronounced /tɑməl/).

`time = "0.1.12"` is the name of the [crate][def-crate] and a [SemVer] version
requirement. The [specifying
dependencies](https://doc.rust-lang.org/cargo/guide/../reference/specifying-dependencies.html) docs have more
information about the options you have here.

If we also wanted to add a dependency on the `regex` crate, we would not need
to add `[dependencies]` for each crate listed. Here's what your whole
`hello_world.rs` file would look like with dependencies on the `time` and `regex`
crates:

```rust
#!/usr/bin/env cargo

//! ```cargo
//! [dependencies]
//! time = "0.1.12"
//! regex = "0.1.41"
//! ```

fn main() {
    let re = Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();
    println!("Did our date match? {}", re.is_match("2014-01-01"));
}
```

You can then re-run this and Cargo will fetch the new dependencies and all of their dependencies.  You can see this by passing in `--verbose`:
```console

> $ cargo eval --verbose ./hello_world.rs

Reviewer 8: Is the `eval` needed here, given that this normally just uses `cargo file.rs`?

Author: Its a leftover from a prior version of this proposal

- `cargo test` and other built-in cargo commands
- `cargo expand` and other third-party cargo commands
- `rust-analyzer` and other editor/IDE integration

While other third-party cargo commands might not immediately adopt single-file
packages, first-party support for them will help encourage their adoption.

This still leaves room for third-party implementations, either differentiating themselves or experimenting with
- Alternative caching mechanisms for lower overhead
- Support for implicit `main`, like doc-comment examples
- Template support for implicit `main` for customizing `use`, `extern`, `#[feature]`, etc
- Short-hand dependency syntax (e.g. `//# serde_json = "*"`)
- Prioritizing other workflows, like runtime performance

# Prior art
[prior-art]: #prior-art

See [Single-file scripts that download their
dependencies](https://dbohdan.com/scripts-with-dependencies)
for a wide view of this space.

Existing Rust solutions:
- [`cargo-script`](https://github.com/DanielKeep/cargo-script)
  - Single-file (`.crs` extension) rust code
    - Partial manifests in a `cargo` doc comment code fence or dependencies in a comment directive

>     - `run-cargo-script` for she-bangs and setting up file associations on Windows

Reviewer 8: ```suggestion
    - `run-cargo-script` for shebangs and setting up file associations on Windows
```

Did our date match? true
```

Cargo will cache the exact information about which revision of all of these dependencies we used.

Now, if `regex` gets updated, we will still build with the same revision until

> we choose to `cargo update --manifest-path hello_world.rs`.

Reviewer 8: How does this work? *Where* is this cached? Where does the equivalent of the lockfile live?

This seems a little too opaque.

Author: That was intentional as lockfiles are another big area for us to try to figure out the right way to handle them is.  The Pre-RFC also has [a section for this](https://github.com/epage/cargo-script-mvs/blob/main/0000-cargo-script.md#lockfile)

Reviewer 8: I'm not suggesting the eRFC has to fully specify embedded lockfiles or any of the many other aspects discussed in the internals thread. I just think if we're going to state that this is cached and updated, we should give *some* idea of whether this is stored in the cargo home directory (which I'd expect), or next to the script (which I hope it isn't), or similar.

(Also, on a separate note, can we make just `cargo update hello_world.rs` work?)

Author: > I'm not suggesting the eRFC has to fully specify embedded lockfiles or any of the many other aspects discussed in the internals thread. I just think if we're going to state that this is cached and updated, we should give some idea of whether this is stored in the cargo home directory (which I'd expect), or next to the script (which I hope it isn't), or similar.

Done

> (Also, on a separate note, can we make just cargo update hello_world.rs work?)

My understanding was that cargo intentionally avoided positional arguments (`cargo add` being an exception) and one reason for this is due to the heavy use of `num_args(0..=1)` arguments like `--timings`.

In the Pre-RFC, I did bring up whether we should provide a short-flag for `--manifest-path` because of how cumbersome it is to use for these files that aren't auto-discovered. 

I'll at least call this out as another area to explore.

Reviewer 10: @epage I just read https://github.com/epage/rfcs/blob/cargo-script/text/3424-cargo-script.md, but couldn't find a place where it mentions what happens with lock files at all. I was searching for `lock` string after reading the whole document.

Author: > Cargo will cache the exact information (in a location referred to as
`CARGO_HOME`) about which revision of all of these dependencies we used.

Author: (I didn't more explicitly call out `Cargo.lock` because, narratively within that document this is "updating", `Cargo.lock` would be meaningless to the user.  That would need to be covered later when talking about multi-file packages)

Reviewer 10: Should we allow some way of getting the information that a Rust developer would expect from a lock file?

Author: What do you mean?

Reviewer 3: > 

isn't that just `cargo metadata --manifest-path=path/to/my-script.rs`?

Reviewer 10: Great. I didn’t use `cargo metadata` before. It will do the trick, albeit in a format that will be unfamiliar for many people.

Reviewer 10: I looked at both `cargo metadata` and `cargo tree` to see if everything contained in a lock file is present. The one thing that I see missing is `checksum` field. I think it can be pretty useful.

Reviewer 10: Turns out that `checksum` is internal implementation checksum useful only to cargo, and is not that useful to end users.

Reviewer 11: One thing I don't see mentioned is whether there's only a single lockfile per system, or one per script. If there's a lockfile per script, what happens if a script is renamed? Does that reset all the dependency versions? What about moving a script file to a different directory without renaming it?

Phrased differently, the document says "Cargo will cache the exact information" but doesn't specify what key the cache is indexed by.

Reviewer 6: That sounds like we'll need `cargo venv` lol
